### PR TITLE
fix issue with now() returning same value across statements

### LINKED ISSUE #3057: once now() is called in a statement, it forever returns the same value
**Describe the bug**

```sql
select now();
-- ... 3 days later ...
select now(); -- returns same old value
```
It's correctly caching now() so it returns the same value when used multiple times in the same query, but it shouldn't be caching the value past statement execution, correct?

**To Reproduce**

**Expected behavior**

Would expect now() to return the same value for all instances in the same query, but a new value for a new statement in the same runtime context.

**Additional context**

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #3057 .

 # Rationale for this change

# What changes are included in this PR?

Root cause of the issue was that down one code path, creating a physical plan wasn't setting the query_execution_start_time in session state. I've fixed it by making the code the same between DataFrame's create_physical_plan() and SessionContext's create_physical_plan(). 

* I added tests to ensure now() returns a consistent value within a statement but different values when called in different statements within the same SessionContext. I added tests for SQL and using the DataFrame API to ensure it works via the different code paths for collecting dataframes.
* views are fixed so that they return now() properly
  * a side-effect of this is that we now store unoptimized plans in view definitions and optimize it on every view execution. I believe this is preferable behavior anyway and also don't see any other way to have now() or functions like that work in views.
* #3069 is partially resolved .. needed to fix the now() expression definition here

I'm still not that happy with how I solved it the root issue with the SessionState cloning via this code duplication, but can follow up with a separate PR.

# Are there any user-facing changes?

now() works correctly across statements in the same session regardless of code path taken to execute the SQL

### CODE REVIEW

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @kmitchener. 

Bonus points if you could put a brief description in the user guide for `now()` while you are here, in the SQL scalar function reference.

**Comment by kmitchener** [comment]
Good thinking, docs updated.

**Comment by kmitchener** [comment]
Actually, I'm going to make this a WIP because I tested with a now() in a view and that code path is still broken. Needs a more robust fix. Will pick it up again in a couple days.

**Comment by liukun4515** [comment]
> Actually, I'm going to make this a WIP because I tested with a now() in a view and that code path is still broken. Needs a more robust fix. Will pick it up again in a couple days.

You can convert this pr as a draft  @kmitchener  If there are some bug or issue not resolved

**Comment by kmitchener** [comment]
@andygrove PTAL

File: datafusion/core/tests/sql/timestamp.rs [hunk 1]
```diff
@@ -436,31 +436,82 @@ async fn test_current_timestamp_expressions() -> Result<()> {
...
-    let t2_naive =
-        chrono::NaiveDateTime::parse_from_str(res1, "%Y-%m-%d %H:%M:%S%.6f").unwrap();
+    Ok(())
+}
 
-    let t2 = t2_naive.timestamp();
-    assert!(t1 <= t2 && t2 <= t3);
-    assert_eq!(res2, res1);
+#[tokio::test]
+async fn test_now_across_statements_using_sql_function() -> Result<()> {
```
Comment (andygrove) [inline datafusion/core/tests/sql/timestamp.rs:454]: It looks like we are missing a SQL test for now() appearing twice in the same statement?

Comment (kmitchener) [reply datafusion/core/tests/sql/timestamp.rs:465]: added

**Review by alamb — COMMENTED** [review]
Thank you for thisPR @kmitchener 👍

**Pull request merged**